### ISSUE TITLE
Outdated imports in docs

### ISSUE THREAD

**Issue by ST-DDT** [SpongePowered/SpongeDocs#640]
I checked the docs for imports that are no longer working with v6.0 and found these:

EDIT: (Removed outdated list)

The test tool to collect all used imports:

````java
public class ImportTest {

    private static Set<String> imports = new TreeSet<>();
    private static int i = 0;

    public static void main(String[] args) {
        File root = new File("../SpongeDocs/source");
        fileStream(root).forEach(ImportTest::test);
        imports.forEach(s -> System.out.println(s + " a" + i++ + " = null;"));
    }

    static Stream<File> fileStream(File file) {
        if (file.isDirectory()) {
            return Arrays.stream(file.listFiles()).flatMap(JavadocRefTest::fileStream);
        } else if (file.isFile() && file.getName().endsWith(".rst")) {
            return Stream.of(file);
        } else {
            return Stream.empty();
        }
    }

    static void test(File file) {
        try {
            Files.lines(file.toPath())
                    .dropWhile(Predicate.isEqual(".. javadoc-import::").negate())
                    .skip(1)
                    .takeWhile(Predicate.isEqual("").negate())
                    .map(String::trim)
                    .forEach(imports::add);
        } catch (IOException e) {}
    }
}
````

**Comment by ST-DDT** [comment]
And these imports are no longer used:

EDIT: (Removed outdated list)

Collected by this tool:

````java
public class JavadocRefTest {

    private static Set<String> unusedImports = new TreeSet<>();
    private static Set<String> missingImports = new TreeSet<>();

    public static void main(String[] args) {
        File root = new File("../SpongeDocs/source");
        fileStream(root).forEach(JavadocRefTest::test);
        unusedImports.forEach(s -> System.out.println(s.substring(root.getPath().length() + 1)));
        missingImports.forEach(s -> System.out.println(s.substring(root.getPath().length() + 1)));
    }

    static Stream<File> fileStream(File file) {
        if (file.isDirectory()) {
            return Arrays.stream(file.listFiles()).flatMap(JavadocRefTest::fileStream);
        } else if (file.isFile() && file.getName().endsWith(".rst")) {
            return Stream.of(file);
        } else {
            return Stream.empty();
        }
    }

    private static final Pattern PATTERN_CLASSNAME = Pattern.compile("([A-Z].+)");
    private static final Pattern PATTERN_JAVADOC = Pattern
            .compile(":javadoc:`([^`#]+)(?>#[A-Za-z0-9_]+(?>\\(([^`]+)?\\))?)?`");

    static void test(File file) {

        Set<String> imports = new TreeSet<>();
        Set<String> references = new TreeSet<>();
        try {
            Files.lines(file.toPath())
                    .dropWhile(Predicate.isEqual(".. javadoc-import::").negate())
                    .skip(1)
                    .takeWhile(Predicate.isEqual("").negate())
                    .map(PATTERN_CLASSNAME::matcher)
                    .flatMap(Matcher::results)
                    .map(r -> r.group(1))
                    .forEach(imports::add);

            Files.lines(file.toPath())
                    .map(PATTERN_JAVADOC::matcher)
                    .flatMap(Matcher::results)
                    .flatMap(r -> Stream.concat(Stream.of(r.group(1).split("<", 2)[0]),
                            Arrays.stream(Objects.toString(r.group(2), "").replace("...", "").split("[,<>] *"))))
                    .filter(s -> !s.isEmpty())
                    .filter(s -> !s.contains("?"))
                    .filter(s -> !s.matches("^[a-z].*"))
                    .filter(s -> s.length() > 1)
                    .forEach(references::add);

        } catch (IOException e) {}
        Set<String> imports2 = new TreeSet<>(imports);

        imports.removeAll(references);
        imports.forEach(r -> unusedImports.add(file.getPath() + " -> " + r));

        references.removeAll(imports2);
        references.forEach(r -> missingImports.add(file.getPath() + " <- " + r));
    }
}
````

**Comment by 12AwsomeMan34** [comment]
As far as I can tell, this doesn't seem to check for outdated method references?
e.g. [Location#removeBlock(Cause)](https://github.com/SpongePowered/SpongeDocs/blob/Update-to-7.0.0/source/plugin/blocks/modifying.rst#changing-a-blocks-type) -> [Location#removeBlock()](https://github.com/SpongePowered/SpongeAPI/blob/bleeding/src/main/java/org/spongepowered/api/world/Location.java#L624)
Since Location itself didn't change, it doesn't appear in the above list.

**Comment by ST-DDT** [comment]
Yeah, I will fix that when I look into this (soon).

**Comment by Inscrutable** [comment]
The list can be whittled down a little, thanks to #656

**Comment by ST-DDT** [comment]
I will try to update the list the weekend.

**Comment by ST-DDT** [comment]
Tested with API v7.0.0 again:

The following imports are no longer valid:

* [x] org.spongepowered.api.data.value.mutable.BoundedValue
* [x] org.spongepowered.api.effect.particle.BlockParticle (#638)
* [ ] org.spongepowered.api.event.cause.entity.spawn.EntitySpawnCause
* [ ] org.spongepowered.api.event.cause.entity.spawn.MobSpawnerSpawnCause
* [ ] org.spongepowered.api.event.cause.entity.spawn.SpawnCause

And these are probably no longer used:

* [x] plugin\data\custom\datamanipulators.rst -> AbstractSingleData
* [x] plugin\data\custom\datamanipulators.rst -> Comparable
* [x] plugin\data\custom\datamanipulators.rst -> ValueContainer
* [x] plugin\data\custom\index.rst -> DataHolder
* [x] plugin\data\custom\index.rst -> DataManipulator
* [x] plugin\data\custom\index.rst -> DataSerializable
* [x] plugin\data\custom\index.rst -> Value
* [x] plugin\data\custom\serialization.rst -> DataManipulator
* [x] plugin\event\causes.rst -> ChangeBlockEvent.Break
* [x] plugin\event\causes.rst -> String
* [x] plugin\event\custom.rst -> Player
* [x] plugin\text\index.rst -> Text

And these are probably missing/wrong

* [x] plugin\commands\childcommands.rst <- CommandElement
* [x] plugin\data\custom\datamanipulators.rst <- String
* [x] plugin\database.rst <- String

@Minecrell : Thanks for pointing out an issue with my test script.

**Comment by ST-DDT** [comment]
There are also references to 
* [ ]  `DataManipulator#registerGettersAndSetters()`
* [ ]  `ImmutableDataManipulator#registerGetters()`
* [ ]  `Cause#get(String, Class)`
* [x] `Location#removeBlock(Cause)`
* [x] `Location#setBlock(BlockState, Cause)`
* [x] `Location#setBlockType(BlockType, Cause)`

that don't seem to be valid any more.

For the location ones I know how to fix them (and PRed it already), any suggestion for the others?

**Comment by ST-DDT** [comment]
Up to this point everything mentioned here has been fixed. 

Except for:

* [x] The imports/references to 
* org.spongepowered.api.event.cause.entity.spawn.EntitySpawnCause
* org.spongepowered.api.event.cause.entity.spawn.MobSpawnerSpawnCause
* org.spongepowered.api.event.cause.entity.spawn.SpawnCause

which were lost in [Split Cause and Context](https://github.com/SpongePowered/SpongeAPI/commit/d5dff5a3431531951b8ff7b37b5b3f56fff235a6#diff-4f1a4edb75f24d729294601d08195996) by @Deamon5550 and @gabizou

It would be nice if those two or any other knowledgeable party could give us a hint on how to update the following two files:

* [Spawning Entities](https://github.com/SpongePowered/SpongeDocs/blob/stable/source/plugin/entities/spawning.rst)
* [Creating Entities](https://github.com/SpongePowered/SpongeDocs/blob/stable/source/plugin/items/creating.rst)

-------------------------------------

* [ ] The same goes for

* org.spongepowered.api.event.cause.Cause#get(String, Class)

in [Event filters](https://github.com/SpongePowered/SpongeDocs/blame/stable/source/plugin/event/filters.rst#L155)

--------------------------------------

* [ ] And also the references to 

* org.spongepowered.api.data.manipulator.DataManipulator#registerGettersAndSetters()
* org.spongepowered.api.data.manipulator.ImmutableDataManipulator#registerGetters()

in

* [Contributing Data Implementations](https://github.com/SpongePowered/SpongeDocs/blob/stable/source/contributing/implementation/datamanipulator.rst)
* [Plugin Custom Data](https://github.com/SpongePowered/SpongeDocs/blob/stable/source/plugin/data/custom/datamanipulators.rst)

**Comment by ST-DDT** [comment]
Improved tooling available in #754
